### PR TITLE
Split PathHierarchy implementation into different files.

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: 

## Summary

This splits the PathHierarchy implementation into separate files for various responsibilities. 

This was previously all in one file to keep as much as possible of its implementation file private. The current length (~1500 lines) is starting to approach the point where the file length makes the code unnecessarily hard to follow. I would have been fine as-is, but some new features and enhancements that I've been working, when added together, start to make this a 2500-3000 line file, which has a real impact on the readability of it.

Splitting this into smaller files meant exposing a _little_ bit of the API, but I find that it still sufficiently encapsulated and that the code readability improvements of spitting into smaller files are worth it.

------

Ignoring the moved—but unchanged—code, these are the _actual_ changes in this PR:

- The access level of the top-level DisambiguationTree struct is changed from `fileprivate` to `internal` so that the other files can use it. It is also:
  - nested within PathHierarchy to indicate that its functionality is specific to that. 
  - renamed to `DisambiguationContainer` to better reflect that its implementation may not always represent a tree.

- The read access level of the`children` property on `PathHierarchy.Node` is changed from `fileprivate` to `internal`. The write access level remains `private(set)`.

- The `availableChildren` data for `PathHierarchy.Error.unknownName` is changed from an array to a set. This makes it consistent with `PathHierarchy.Error.notFound` and better reflect that the order doesn't matter because the values are sorted by near-miss proximity to the next path component in the authored link.

## Dependencies

None. 
## Testing

None. This isn't a user-facing change.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- ~[ ] Added tests~
- [x] Ran the `./bin/test` script and it succeeded
- ~[ ] Updated documentation if necessary~

### CODE REVIEW

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Review by ethan-kusters — APPROVED** [review]
